File: react-native/storyly-placement-react-native/ios/Common/Data/Widgets/BannerData.swift
import Foundation
import StorylyCore
import StorylyBanner

// MARK: - Banner Data Payload Encoding

func encodeBannerDataPayload(_ data: BannerDataPayload) -> [String: Any] {
    return [
        "type": STRWidgetType.banner.description,
        "items": data.items.map { encodeBannerSlide($0) }.compactMap { $0 }
    ]
}

func encodeBannerPayload(_ payload: BannerPayload?) -> [String: Any]? {
    guard let payload = payload else { return nil }
    
    return [
        "component": encodeBannerComponent(payload.component),
        "item": encodeBannerSlide(payload.item)
    ].compactMapValues { $0 }
}

// MARK: - Banner Slide Encoding

func encodeBannerSlide(_ slide: BannerSlide?) -> [String: Any]? {
    guard let slide = slide else { return nil }
    
    return ([
        "id": slide.uniqueId,
        "name": slide.name,
        "index": slide.index,
        "actionUrl": slide.actionUrl,
        "componentList": slide.componentList?.map { encodeBannerComponent($0) }.compactMap { $0 },
        "actionProducts": slide.actionProducts?.map { encodeSTRProductItem($0) },
        "currentTime": slide.currentTime
    ]  as [String: Any?]).compactMapValues { $0 }
}

// MARK: - Banner Component Encoding

func encodeBannerComponent(_ component: BannerComponent?) -> [String: Any]? {
    guard let component = component else { return nil }
    
    var result: [String: Any] = [
        "id": component.id,
        "type": component.type.get(),
        "customPayload": component.customPayload as Any
    ]
    
    if let button = component as? BannerButtonComponent {
        result["text"] = button.text
        result["actionUrl"] = button.actionUrl
        result["products"] = button.products?.map { encodeSTRProductItem($0) }
    }
    
    return result.compactMapValues { $0 }
}

File: placement-flutter/storyly_placement_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_placement_flutter/common/data/widgets/BannerData.kt
package com.appsamurai.storyly.storyly_placement_flutter.common.data.widgets

import com.appsamurai.storyly.banner.data.BannerButtonComponent
import com.appsamurai.storyly.banner.data.BannerComponent
import com.appsamurai.storyly.banner.data.BannerDataPayload
import com.appsamurai.storyly.banner.data.BannerSlide
import com.appsamurai.storyly.banner.data.model.BannerPayload
import com.appsamurai.storyly.core.STRWidgetType
import com.appsamurai.storyly.core.data.model.product.STRProductItem
import com.appsamurai.storyly.storybar.data.model.StoryBarPayload
import com.appsamurai.storyly.storybar.ui.model.StoryButtonComponent
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.encodeSTRProductItem

fun encodeBannerDataPayload(data: BannerDataPayload): Map<String, Any?> {
    return mapOf(
        "type" to STRWidgetType.Banner.raw,
        "items" to data.items.map { encodeBannerSlide(it) }
    )
}


fun encodeBannerPayload(payload: BannerPayload?): Map<String, Any?>? {
    payload ?: return null
    return mapOf(
        "component" to encodeBannerComponent(payload.component),
        "item" to encodeBannerSlide(payload.item),
    )
}


fun encodeBannerSlide(slide: BannerSlide?): Map<String, Any?>? {
    slide ?: return null
    return mapOf(
        "id" to slide.uniqueId,
        "name" to slide.name,
        "index" to slide.index,
        "actionUrl" to slide.actionUrl,
        "componentList" to slide.componentList?.map { encodeBannerComponent(it) },
        "actionProducts" to slide.actionProducts?.map { encodeSTRProductItem(it) },
        "currentTime" to slide.currentTime
    )
}


fun encodeBannerComponent(component: BannerComponent?): Map<String, Any?>? {
    component ?: return null
    val base = mapOf(
        "id" to component.id,
        "type" to component.type.name.lowercase(),
        "customPayload" to component.customPayload
    )
    val additional = when (component) {
        is BannerButtonComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        else -> emptyMap()
    }
    return base + additional
}
